mkosi/__init__.py

```diff
@@ -543,6 +551,234 @@ def install_systemd_boot(state: MkosiState) -> None:
+        return False
+
+    have = find_grub_bios_directory(state) is not None
+    if not have and state.config.bootable == ConfigFeature.enabled:
+        die("A BIOS bootable image with grub was requested but grub for BIOS is not installed")
+
+    bios = any(p.type == Partition.GRUB_BOOT_PARTITION_UUID for p in partitions)
+    if partitions and not bios and state.config.bootable == ConfigFeature.enabled:
+        die("A BIOS bootable image with grub was requested but no BIOS Boot Partition was configured")
+
+    root = any(p.type.startswith("root") or p.type.startswith("usr") for p in partitions)
+    if partitions and not root and state.config.bootable == ConfigFeature.enabled:
+        die("A BIOS bootable image with grub was requested but no root or usr partition was configured")
+
+    install = find_grub_install(state) is not None
+    if not install and state.config.bootable == ConfigFeature.enabled:
+        die("A BIOS bootable image with grub was requested but grub-install was not found")
+
+    want = (have and bios and root and install) if partitions else have
+
+    control = Path("/dev/loop-control")
+    if not control.exists():
+        if state.config.bootable == ConfigFeature.enabled:
+            die("A BIOS bootable image with grub was requested but grub cannot be installed because /dev/loop-control is not available")
+        elif want:
+            logging.warning("Cannot install grub because /dev/loop-control is missing, skipping")
+
+        return False
+    elif control.stat().st_uid != os.getuid():
+        if state.config.bootable == ConfigFeature.enabled:
+            die(
+                textwrap.dedent(
+                    """\
+                    A BIOS bootable image with grub was requested but grub cannot be installed because
+                    mkosi is running in a user namespace. Run mkosi with root privileges outside of any
+                    unprivileged containers to solve this problem.
+                    """
+                )
+            )
+        elif want:
+            logging.warning("Cannot install grub because mkosi is running in a user namespace, skipping")
+
+        return False
+
+    return want
+
+
+def prepare_grub_config(state: MkosiState) -> Optional[Path]:
+    install = find_grub_install(state)
+    if not install:
+        return None
+
+    prefix = "grub2" if "grub2" in os.fspath(install) else "grub"
+
+    config = state.root / "efi" / prefix / "grub.cfg"
+    with umask(~0o700):
+        config.parent.mkdir(exist_ok=True)
+
+    # For some unknown reason, if we don't set the timeout to zero, grub never leaves its menu, so we default
+    # to a zero timeout, but only if the config file hasn't been provided by the user.
+    if not config.exists():
+        with umask(~0o600), config.open("w") as f:
+            f.write("set timeout=0\n")
+
+    # EFI grub shipped by distributions reads its configuration from /efi/<distribution>/grub.cfg so let's
```

Comment:
```suggestion
    # EFI grub shipped by distributions reads its configuration from $ESP/EFI/<distribution>/grub.cfg so let's
```
would be more in line with the definition of `efi` below